File: DSA/Khushi_590028673/Day18/Q2.cpp
#include <iostream>
#include <vector>
using namespace std;

// Definition for singly-linked list
struct ListNode {
    int val;
    ListNode *next;

    ListNode(int x) {
        val = x;
        next = NULL;
    }
};

// Create Linked List
ListNode* createList(vector<int> arr) {
    if (arr.empty()) return NULL;

    ListNode* head = new ListNode(arr[0]);
    ListNode* temp = head;

    for (int i = 1; i < arr.size(); i++) {
        temp->next = new ListNode(arr[i]);
        temp = temp->next;
    }

    return head;
}

// Print Linked List
void printList(ListNode* head) {
    if (head == NULL) {
        cout << "[]";
        return;
    }

    while (head) {
        cout << head->val;
        if (head->next)
            cout << " -> ";
        head = head->next;
    }
    cout << endl;
}

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {

        ListNode* dummy = new ListNode(0);
        dummy->next = head;

        ListNode* fast = dummy;
        ListNode* slow = dummy;

        // Move fast pointer n+1 steps ahead
        for (int i = 0; i <= n; i++) {
            fast = fast->next;
        }

        // Move both pointers
        while (fast != NULL) {
            fast = fast->next;
            slow = slow->next;
        }

        // Delete nth node from end
        ListNode* delNode = slow->next;
        slow->next = delNode->next;
        delete delNode;

        ListNode* newHead = dummy->next;
        delete dummy;

        return newHead;
    }
};

int main() {

    int n;
    cout << "Enter number of nodes: ";
    cin >> n;

    vector<int> arr(n);

    cout << "Enter elements: ";
    for (int i = 0; i < n; i++)
        cin >> arr[i];

    int k;
    cout << "Enter n (node to remove from end): ";
    cin >> k;

    ListNode* head = createList(arr);

    cout << "Original List: ";
    printList(head);

    Solution obj;
    head = obj.removeNthFromEnd(head, k);

    cout << "Updated List: ";
    printList(head);

    return 0;
}

File: DSA/NeeravKumarSharma_590018263/Day_18/Q2.java
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Q2 {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode fast = dummy;
        ListNode slow = dummy;

        for (int i = 0; i <= n; i++) {
            fast = fast.next;
        }

        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }

        slow.next = slow.next.next;
        return dummy.next;
    }

    public static void main(String[] args) {
        ListNode list = new ListNode(1,new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5,null)))));

        Q2 obj = new Q2();
        
        int n = 2;
        ListNode updatedHead = obj.removeNthFromEnd(list, n);
        
        System.out.print("List after removing the node" + n + " node before end: ");
        ListNode currentNode = updatedHead;
        while (currentNode != null) {
            System.out.print(currentNode.val + (currentNode.next != null ? " -> " : ""));
            currentNode = currentNode.next;
        }
        System.out.println();
    }
}